app/src/test/java/com/kenchen/capstonenewsapp/NewsRepositoryImpTest.kt

```diff
@@ -0,0 +1,198 @@
+        }
+    }
+
+    @Test
+    fun `When articles are searched through title and no articles are matched return empty list`() {
+        val expectedResult = 0
+
+        coEvery { mockArticleDao.searchArticles(any()) } returns listOf()
+
+        runBlocking {
+            val articles = newsRepository.searchArticles("US")
+
+            assertEquals(expectedResult, articles.size)
+        }
+    }
+
+    @Test
+    fun `toggle data usage`() {
+        runBlocking {
+            coEvery { mockPrefsStore.toggleDataUsage() } just runs
+            newsRepository.toggleDataUsage()
+            coVerify(exactly = 1) { mockPrefsStore.toggleDataUsage() }
+        }
+    }
+
+    @Test
+    fun `is data usage should returns flow of boolean`() {
+        val expectedResult = true
+
+        runBlocking {
+            coEvery { mockPrefsStore.isDataUsage() } returns flow { emit(true) }
+
+            val actualResult = newsRepository.isDataUsage().first()
+
+            assertEquals(expectedResult, actualResult)
+        }
+    }
+
+    @Test
+    fun `articles are fetched from database first and ArticleState ready with articles is returned`() {
+        val expectedResult = ArticleState.Ready(listOf(dummyArticle))
+
+        coEvery { mockArticleDao.getArticles() } returns listOf(dummyArticle)
+
+        val flowArticleState = newsRepository.getArticles()
+
+        runBlocking {
+            assertEquals(expectedResult, flowArticleState.first())
+        }
+    }
+
+//    @Test
+//    fun `fetch with wifi enabled, remote api success ArticleState ready is returned`() {
+//        val expectedException = Exception("error message")
+//        every { mockPrefsStore.isDataUsage() } returns flow { emit(false) }
+//        every { mockNetworkStatusChecker.isOnWifiConnection() } returns true
+//        coEvery { mockRemoteApi.getTopHeadlinesByCountry("us") } returns listOf(dummyArticle)
+//        coEvery { mockArticleDao.getArticles() } returns listOf(dummyArticle)
+//
+//        val flowArticleState = newsRepository.getArticles()
+//
+//        runBlocking {
+//            assertEquals(ArticleState.Ready(listOf(dummyArticle)),
+//                flowArticleState.drop(1).first()
+//            )
+//        }
+//    }
+
+    @Test
+    fun `fetch with wifi enabled, remote api success return ArticleState partial with articles`() {
+        coEvery { mockArticleDao.getArticles() } returns listOf(dummyArticle)
+        every { mockPrefsStore.isDataUsage() } returns flow { emit(true) }
+        every { mockNetworkStatusChecker.isOnWifiConnection() } returns true
+        coEvery { mockRemoteApi.getTopHeadlinesByCountry("us") } returns listOf(dummyArticle)
+
+        coEvery { mockArticleDao.clearArticles() } just runs
+        coEvery { mockArticleDao.addArticles(any()) } just runs
+
+        runBlocking {
+            val flowArticleState = newsRepository.getArticles()
+
+//            println(flowArticleState.take(3).toList())
```

Comment:
I never did get a chance to look in to this issue with `take()` and `drop()` 😅 Overall though, you have really good tests here for your repo that do a pretty thorough job of testing different edge cases!